src/options.ts

```diff
@@ -0,0 +1,279 @@
+  waitQueueMultiple: URIType.number,
+  waitQueueTimeoutMS: URIType.number,
+  readConcernLevel: URIType.string,
+  readPreference: URIType.string,
+  maxStalenessSeconds: URIType.number,
+  readPreferenceTags: URIType.stringArray,
+  authSource: URIType.string,
+  authMechanism: URIType.string,
+  authMechanismProperties: {
+    SERVICE_NAME: URIType.string,
+    CANONICALIZE_HOST_NAME: URIType.boolean,
+    SERVICE_REALM: URIType.string,
+    AWS_SESSION_TOKEN: URIType.string
+  },
+  gssapiServiceName: URIType.string,
+  localThresholdMS: URIType.number,
+  serverSelectionTimeoutMS: URIType.number,
+  serverSelectionTryOnce: URIType.boolean,
+  heartbeatFrequencyMS: URIType.number,
+  appName: URIType.string,
+  retryReads: URIType.boolean,
+  retryWrites: URIType.boolean,
+  directConnection: URIType.boolean,
+  // write concern options
+  w: URIType.boolean,
+  wtimeoutMS: URIType.number,
+  j: URIType.boolean,
+  fsync: URIType.boolean
+} as const;
+
+type URIObject = { [key: string]: URIType | URIObject };
+
+type StringObject = { [key: string]: any };
+type URIValue<T> = T extends URIType.string
+  ? string
+  : T extends URIType.number
+  ? number
+  : T extends URIType.boolean
+  ? boolean
+  : T extends URIType.stringArray
+  ? string[]
+  : T extends object
+  ? Partial<{ [K in keyof T]: URIValue<T[K]> }>
+  : never;
+
+type Lookup = { [key: string]: { normalized: string; type: URIType | Lookup } };
+
+type ObjectKeys<T> = T extends object
+  ? (keyof T)[]
+  : T extends number
+  ? []
+  : T extends Array<any> | string
+  ? string[]
+  : never;
+
+function ObjectKeys<T>(o: T): ObjectKeys<T> {
+  return Object.keys(o) as ObjectKeys<T>;
+}
+
+/**
+ * Transforms query string object types from MongoURIOptions lookup, this does not validate the options.
+ */
+function typeURIOptions(input: QS.ParsedUrlQuery): URIValue<typeof MongoURIOptions> {
+  function lookupRecursive(obj: URIObject): Lookup {
+    const keys = ObjectKeys(obj);
+    return keys.reduce((types, normalized) => {
+      let type: URIType | URIObject | Lookup = obj[normalized];
+      const keyLower = normalized.toString().toLowerCase();
+      if (typeof type === 'object') {
+        type = lookupRecursive(type);
+      }
+      return { ...types, [keyLower]: { normalized, type } };
+    }, {});
+  }
+
+  const lookup = lookupRecursive(MongoURIOptions);
+
+  const options = Object.keys(input).reduce((options, key) => {
+    const keyLower = key.toLowerCase();
+    let value = input[key];
+    const ref = lookup[keyLower];
+    if (!ref) {
+      console.warn(`unknown property ${key}`);
+      return options;
+    }
+    const { type, normalized } = ref;
+    let currentValue = options[normalized as keyof typeof MongoURIOptions];
+    let results;
+
+    function recursive(
+      normalized: string,
+      value: string | string[] | undefined,
+      type: URIType | Lookup
+    ): URIValue<typeof MongoURIOptions> {
+      const match = (t: URIType) => type === t;
```

Comment:
Very ML approach here with the match (which is cool), is there advantage over a switch block?

Reply:
I think I did originally do a switch, and I switched it (pun-intended) I'm a little switch-adverse, and I know Mr. Broadstone is as well.